lib/teleterm/gateway/gateway.go

```diff
@@ -116,13 +141,90 @@ func (g *Gateway) Close() error {
 
 // Serve starts the underlying ALPN proxy. Blocks until closeContext is canceled.
 func (g *Gateway) Serve() error {
-	g.Log.Info("Gateway is open.")
+	g.cfg.Log.Info("Gateway is open.")
 
 	if err := g.localProxy.Start(g.closeContext); err != nil {
 		return trace.Wrap(err)
 	}
 
-	g.Log.Info("Gateway has closed.")
+	g.cfg.Log.Info("Gateway has closed.")
+
+	return nil
+}
+
+func (g *Gateway) URI() uri.ResourceURI {
+	return g.cfg.URI
+}
+
+func (g *Gateway) SetURI(newURI uri.ResourceURI) {
+	g.cfg.URI = newURI
+}
+
+func (g *Gateway) TargetURI() string {
+	return g.cfg.TargetURI
+}
+
+func (g *Gateway) TargetName() string {
+	return g.cfg.TargetName
+}
+
+func (g *Gateway) Protocol() string {
+	return g.cfg.Protocol
+}
+
+func (g *Gateway) TargetUser() string {
+	return g.cfg.TargetUser
+}
+
+func (g *Gateway) TargetSubresourceName() string {
+	return g.cfg.TargetSubresourceName
+}
+
+func (g *Gateway) SetTargetSubresourceName(value string) {
+	g.cfg.TargetSubresourceName = value
+}
+
+func (g *Gateway) Log() *logrus.Entry {
+	return g.cfg.Log
+}
+
+func (g *Gateway) LocalAddress() string {
+	return g.cfg.LocalAddress
+}
+
+func (g *Gateway) LocalPort() string {
+	return g.cfg.LocalPort
+}
+
+// SetLocalPort attempts to create a listener on the specified port. If successful, it stops
+// g.localProxy and starts a new one with the new listener and then updates the fields on gateway.
+//
+// If it fails it's imperative that the current proxy is kept intact and the fields on gateway are
+// not changed. This way if the user attempts to change the port to one that cannot be obtained,
+// they're able to correct that mistake and choose a different port.
+func (g *Gateway) SetLocalPort(port string) error {
+	result, err := newListenerAndLocalProxy(*g.cfg, port)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
+	if err = g.localProxy.Close(); err != nil {
+		g.cfg.Log.WithError(err).Warnf("Failed to close the previous proxy on port %s.", previousPort)
+	}
+
+	g.cfg.LocalPort = result.LocalPort
+	g.localProxy = result.localProxy
+	g.closeCancel = result.closeCancel
+	g.closeContext = result.closeContext
+
+	go func() {
+		if err := g.Serve(); err != nil {
+			g.cfg.Log.WithError(err).Warn("Failed to restart the gateway after changing the port.")
+		}
+	}()
```

Comment:
Any operation on gateways must go through `daemon.Service` methods which obtain a write lock before making any modifications, so I think we're protected against that. Well, unless we write code within `gateway` which calls those functions twice.

If `g.Serve` is never called before `SetLocalPort` then I think everything should be fine? The context will be closed which will close the listener and then `Close` on local proxy will be a noop I think.

Is it worthwhile to add another lock on the `gateway.Gateway` layer that would be called in setters?

Reply:
> Any operation on gateways must go through daemon.Service methods which obtain a write lock before making any modifications

Ah, I missed the lock part on the outside. Maybe let's add a comment somewhere how they are protected.

> If g.Serve is never called before SetLocalPort then I think everything should be fine? The context will be closed which will close the listener and then Close on local proxy will be a noop I think.

Yes, closing the not-started local proxy should be fine. Though, I am not sure if there is any side effect if `g.Serve` is called twice on the same proxy. Probably ok too. But more importantly, if `SetLocalPort` is called first, the caller may not expect the local proxy to start until the caller decides to call `Serve`. If there is no easy way to address this, maybe leave a comment `SetLocalPort` is expected to be called after `Serve`. What do you want?

Reply:
> Ah, I missed the lock part on the outside. Maybe let's add a comment somewhere how they are protected.

Sure.

> But more importantly, if `SetLocalPort` is called first, the caller may not expect the local proxy to start until the caller decides to call `Serve`. If there is no easy way to address this, maybe leave a comment `SetLocalPort` is expected to be called after `Serve`. What do you think?

I suppose one way to solve this would be to add a private field called sth like `serving` and then not do the equivalent of `Serve` if `serving` is true. But even then what would irk me the most is that `SetLocalPort` pretends to be a simple setter. In reality, there's more logic involved which just has to be contained somewhere.

I think for now I'll rename `SetLocalPort` to `SetLocalPortAndRestart` and add a comment saying that `SetLocalPortAndRestart` is expected to be called after `Serve`.